### PR TITLE
Move BPF event filtering to the kernel

### PR DESCRIPTION
**Pull request by jakule**
https://github.com/gravitational/teleport/pull/19354 moved filtering of disk events to the kernel space. This PR continues work in this area and moves all events to be filtered in the kernel space.

### CODE REVIEW

**Comment by jakule** [comment]
@strideynet @greedy52 Friendly ping.

File: bpf/enhancedrecording/command.bpf.c [hunk 1]
```diff
@@ -61,9 +65,18 @@ static int enter_execve(const char *filename,
     // create data here and pass to submit_arg to save stack space (#555)
     struct data_t data = {};
     struct task_struct *task;
+    u64 cgroup = bpf_get_current_cgroup_id();
+    u64 *is_monitored;
+
+    // Check if the cgroup should be monitored.
+    is_monitored = bpf_map_lookup_elem(&monitored_cgroups, &cgroup);
+    if (is_monitored == NULL) {
+        // Missed entry.
```
Comment (strideynet) [inline bpf/enhancedrecording/command.bpf.c:74]:
```suggestion
        // cgroup has not been marked for monitoring, so ignore.
```

**Review by strideynet — APPROVED** [review]
I can't claim to be intimately familiar with eBPF, but, I'm pretty happy with this though.

File: bpf/enhancedrecording/common.h [hunk 2]
```diff
@@ -0,0 +1,3 @@
+
+// Maximum monitored sessions.
+#define MAX_MONITORED_SESSIONS 1024
```
Comment (greedy52) [inline bpf/enhancedrecording/common.h:3]: Should we add `#ifndef` for header here just in case?

File: bpf/enhancedrecording/network.bpf.c [hunk 3]
```diff
@@ -55,7 +59,16 @@ static int trace_connect_entry(struct sock *sk)
 static int trace_connect_return(int ret, short ipver)
 {
     u64 pid_tgid = bpf_get_current_pid_tgid();
-	u32 id = (u32)pid_tgid;
+    u64 cgroup = bpf_get_current_cgroup_id();
+	  u32 id = (u32)pid_tgid;
+	  u64 *is_monitored;
```
Comment (greedy52) [inline bpf/enhancedrecording/network.bpf.c:64]: nit: seems some mixed tabs and spaces here.

File: lib/bpf/bpf.go [hunk 4]
```diff
@@ -233,9 +233,15 @@ func (s *Service) OpenSession(ctx *SessionContext) (uint64, error) {
...
-		return 0, trace.Wrap(err)
+	for _, module := range []interface{ startSession(cgroupID uint64) error }{
+		s.open,
+		s.exec,
+		s.conn,
+	} {
+		// Register cgroup in the BPF module.
+		if err := module.startSession(cgroupID); err != nil {
+			return 0, trace.Wrap(err)
+		}
```
Comment (greedy52) [inline lib/bpf/bpf.go:244]: not very familiar with bpf. but if say the second or the third in the loop fails, do we have to cleanup the earlier ones?

Comment (jakule) [reply lib/bpf/bpf.go]: Good point

File: lib/bpf/common_test.go [hunk 5]
```diff
@@ -101,9 +120,32 @@ func TestCheckAndSetDefaults(t *testing.T) {
...
+// callEndpoint wait for continue signal to be generated an executes HTTP GET
+// on provided endpoint.
+func callEndpoint(endpoint string) error {
+	if err := waitForContinue(); err != nil {
+		return err
+	}
+
+	resp, err := http.Get(endpoint)
+	// Close the body to make our linter happy.
+	_ = resp.Body.Close()
```
[the remark this answers is not in the document]
Comment (jakule) [reply lib/bpf/common_test.go]: In this really doesn't matter, but if we want to be correct let's do it http://devs.cloudimmunity.com/gotchas-and-common-mistakes-in-go-golang/index.html#close_http_resp_body

File: bpf/enhancedrecording/common.h (hunk 2 above)
Comment (jakule) [reply bpf/enhancedrecording/common.h]: Added

File: bpf/enhancedrecording/network.bpf.c (hunk 3 above)
Comment (jakule) [reply bpf/enhancedrecording/network.bpf.c]: Fixed

File: bpf/enhancedrecording/command.bpf.c (hunk 1 above)
Comment (jakule) [reply bpf/enhancedrecording/command.bpf.c:74]: Changed in all places

**Review by greedy52 — APPROVED** [review]

**Pull request merged**